### PR TITLE
Causality trait support for window functions

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

### CODE REVIEW

**Review by metesynnada — APPROVED** [review]
LGTM!

File: datafusion/physical-expr/src/window/built_in.rs [hunk 1]
```diff
@@ -206,7 +206,12 @@ impl WindowExpr for BuiltInWindowExpr {
             let record_batch = &partition_batch_state.record_batch;
             let num_rows = record_batch.num_rows();
             let mut row_wise_results: Vec<ScalarValue> = vec![];
-            let mut is_causal = self.window_frame.is_causal();
+            // let mut is_causal = self.window_frame.is_causal();
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/built_in.rs:209]: I think you forgot to remove this

File: datafusion/physical-expr/src/window/lead_lag.rs [hunk 2]
```diff
@@ -195,6 +195,11 @@ impl PartitionEvaluator for WindowShiftEvaluator {
         }
     }
 
+    fn is_causal(&self) -> bool {
+        // if window shift is lagging we do not need any future rows
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/lead_lag.rs:199]:
```suggestion
        // Lagging windows are causal by definition:
```

File: datafusion/physical-expr/src/window/row_number.rs [hunk 3]
```diff
@@ -92,6 +92,11 @@ pub(crate) struct NumRowsEvaluator {
 }
 
 impl PartitionEvaluator for NumRowsEvaluator {
+    fn is_causal(&self) -> bool {
+        // row number doesn't need any future value
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/row_number.rs:96]:
```suggestion
        // The ROW_NUMBER function doesn't need "future" values to emit results:
```

**Review by ozankabak — APPROVED** [review]
Few minor changes, LGTM otherwise. Please send upstream after applying.

**Comment by ozankabak** [comment]
Merged upstream.

**Pull request closed without merging**